fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/TextMatchBasedRewriteRule.java

```diff
@@ -0,0 +1,366 @@
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptExpressionVisitor;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.OptimizerTraceUtil;
+import com.starrocks.sql.optimizer.operator.Operator;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalOlapScanOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalTopNOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rule.Rule;
+import com.starrocks.sql.optimizer.rule.RuleType;
+import com.starrocks.sql.optimizer.rule.transformation.materialization.MvUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.starrocks.sql.optimizer.OptimizerTraceUtil.logMVRewrite;
+import static com.starrocks.sql.optimizer.rule.transformation.materialization.MaterializedViewRewriter.REWRITE_SUCCESS;
+
+/**
+ * A rule that performs text match based rewrite on a query using materialized views.
+ */
+public class TextMatchBasedRewriteRule extends Rule {
+    private static final Logger LOG = LogManager.getLogger(TextMatchBasedRewriteRule.class);
+
+    private final ConnectContext connectContext;
+    private final StatementBase stmt;
+    private final Map<Operator, ParseNode> optToAstMap;
+
+    // To avoid text match costing too much time, use parameters below to limit it.
+    // limit for sub-query text match(default 4), no match when it <= 0
+    private final int mvSubQueryTextMatchMaxCount;
+    // limit for mvs which matched input query(default 64)
+    private final long mvRewriteRelatedMVsLimit;
+
+    private int subQueryTextMatchCount = 1;
+
+    public TextMatchBasedRewriteRule(ConnectContext connectContext,
+                                     StatementBase stmt,
+                                     Map<Operator, ParseNode> optToAstMap) {
+        super(RuleType.TF_MV_TEXT_MATCH_REWRITE_RULE, Pattern.create(OperatorType.PATTERN));
+
+        this.connectContext = connectContext;
+        this.stmt = stmt;
+        this.optToAstMap = optToAstMap;
+        this.mvSubQueryTextMatchMaxCount =
+                connectContext.getSessionVariable().getMaterializedViewSubQueryTextMatchMaxCount();
+        this.mvRewriteRelatedMVsLimit =
+                connectContext.getSessionVariable().getCboMaterializedViewRewriteRelatedMVsLimit();
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        OptExpression rewritten = doTransform(context, input, stmt);
```

Comment:
why not use the framework of ruleRewriteIterative?
seems that you repeat the thing of ruleRewriteIterative and RewriteTreeTask?

Reply:
just like `MaterializedViewRule` , it's same to implement. and it's no need to call `ruleRewriteIterative ` since there is no `rootTaskContext` at this moment.